src/lib/math/pcurves/pcurves.h

```diff
@@ -0,0 +1,300 @@
+/*
+* (C) 2024 Jack Lloyd
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#ifndef BOTAN_PCURVES_H_
+#define BOTAN_PCURVES_H_
+
+#include <botan/secmem.h>
+#include <botan/types.h>
+#include <array>
+#include <optional>
+#include <span>
+#include <string_view>
+#include <vector>
+
+namespace Botan {
+
+class RandomNumberGenerator;
+
+#if defined(BOTAN_HAS_ASN1)
+class OID;
+#endif
+
+}  // namespace Botan
+
+namespace Botan::PCurve {
+
+/// Identifier for a named prime order curve
+class BOTAN_TEST_API PrimeOrderCurveId {
+   public:
+      enum class Id : uint8_t {
+         /// secp256r1 aka P-256
+         secp256r1,
+         /// secp384r1 aka P-384
+         secp384r1,
+         /// secp521r1 aka P-521
+         secp521r1,
+         /// secp256k1
+         secp256k1,
+         /// brainpool256r1
+         brainpool256r1,
+         brainpool384r1,
+         brainpool512r1,
+         frp256v1,
+         sm2p256v1,
+      };
+
+      using enum Id;
+
+      Id code() const { return m_id; }
+
+      static std::vector<PrimeOrderCurveId> all();
+
+      std::string to_string() const;
+
+      PrimeOrderCurveId(Id id) : m_id(id) {}
+
+      /// Map a string to a curve identifier
+      static std::optional<PrimeOrderCurveId> from_string(std::string_view name);
+
+#if defined(BOTAN_HAS_ASN1)
+      /// Map an OID to a curve identifier
+      ///
+      /// Uses the internal OID table
+      static std::optional<PrimeOrderCurveId> from_oid(const OID& oid);
+#endif
+
+   private:
+      const Id m_id;
+};
+
+class BOTAN_TEST_API PrimeOrderCurve {
+   public:
+      /// Somewhat arbitrary maximum size for a field or scalar
+      ///
+      /// Sized to fit at least P-521
+      static const size_t MaximumBitLength = 521;
+
+      static const size_t MaximumByteLength = (MaximumBitLength + 7) / 8;
+
+      /// Number of words used to store MaximumByteLength
+      static const size_t StorageWords = (MaximumByteLength + sizeof(word) - 1) / sizeof(word);
+
+      static std::shared_ptr<const PrimeOrderCurve> from_name(std::string_view name) {
+         if(auto id = PrimeOrderCurveId::from_string(name)) {
+            return PrimeOrderCurve::from_id(id.value());
+         } else {
+            return {};
+         }
+      }
+
+      static std::shared_ptr<const PrimeOrderCurve> from_id(PrimeOrderCurveId id);
```

Comment:
> Though, it does pose a limitation for potential (internal) use cases where the required curve is known at compile time.

I don't think we have many scenarios like that, at least at the moment.

It's certainly something we could do but I don't see a reason to do it preemptively lacking a clear use.

> This could (and perhaps should) be for internal use only

Keep in mind that none of this is or will be exposed to users. The public facing side of this is #4042